Feature/float8

maddyscientist opened the pull request:
This PR adds support for a new data order, `QUDA_FLOAT8_FIELD_ORDER`, for the fermion fields and `QUDA_FLOAT8_GAUGE_FIELD_ORDER` for gauge fields
* Verified from generated assembly code that 128-bit load/store transactions are used for both the `colorspinor::FloatNOrder` accessor and for the blas kernels.
* Since this is an experimental option, must be opted into with cmake advanced option `QUDA_FLOAT8=ON`
* When enabled, half and quarter precision blas and copy kernels will be instantiated
* When enabled, half and quarter precision with either recon-8 or recon-9 gauge fields will use float-8 ordering as their native data order
* Robustness additions to accessor framework to allow float-8 ordered fields to work even when the (spin-projected) halo is not compatible with float-8 ordering - use float-4 ordering for the halo, and force use of halo-packing kernel in the temporal direction
* While I have not added this to this PR, I have verified that Wilson dslash, including multi-GPU, with float-8 ordering just works
* float-8 ordering is only enabled when `QUDA_TEX=OFF`.  Compile time check added to ensure that this is the case
* Fix some bugs introduced with #942

maddyscientist committed:
Add initial support for float8-ordered fields.  For now, only instantiated for half and quarter precision blas

maddyscientist committed:
Cleanup of pre-dslash order, precision and location checking

maddyscientist committed:
Tidy up of dwf-5d packing handling

maddyscientist committed:
colorspinor::FloatNOrder is now robust to handle float-8 ordering: 1.) static_assert to check we have a valid data order, and 2.) if short-vector isn't applicable to spin-projected halo halve the short-vector length

maddyscientist committed:
If data order for halo is different than for bulk, we must use a packing kernel for the halo

maddyscientist committed:
Add QUDA_FLOAT8 to cmake options to enable float8-ordered blas for half and quarter precision.  Aded check to ensure that QUDA_TEX is not enabled with QUDA_FLOAT8

maddyscientist committed:
Fix bug in last commit

maddyscientist committed:
Add short8 and char8 specializations to vector_load to ensure loads are vectorized

maddyscientist committed:
Add QUDA_FLOAT8_GAUGE_ORDER for fixed-point recon-8/recon-9 fields.  Enabled only when QUDA_FLOAT8=ON

maddyscientist committed:
Cleanup and fix some issues from #942

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/float8

maddyscientist committed:
Quick fix for CG solver to allow it to be embedded as a preconditioner

maddyscientist committed:
Redefine offset in FloatNOrder accessors to include vector size.  Remove now redundant tex_offset.

maddyscientist committed:
Rework of vector_load/vector_store helpers to allow for potential optimization of indexing

maddyscientist committed:
Fix bug in last commit

maddyscientist committed:
Optimization of ColorSpinor::reconstruct to remove multiplication by zero (which isn't optimized out by the compiler)

maddyscientist committed:
Optimization of ColorSpinor::project to remove multiplication by zero (which isn't optimized out by the compiler)

maddyscientist committed:
slight improvement to recon-13 and recon-9

mathiaswagner commented on lib/reduce_quda.cu:
```diff
@@ -440,6 +440,15 @@ namespace quda {
+                = nativeReduce<doubleN, ReduceType, float8, short8, short8, M, Reducer, writeX, writeY, writeZ, writeW, writeV>(
+                    a, b, x, y, z, w, v, y.Volume());
+#else
+            errorQuda("blas has not been built for Nspin=%d fields", x.Nspin());
```
should we include the FieldOrder in this error message?

mathiaswagner commented on include/float_vector.h:
```diff
@@ -45,7 +52,7 @@ namespace quda {
   }
 
-  __host__ __device__ inline float4 operator*(const float a, const float4 x) {
+  __host__ __device__ inline float4 operator*(const float &a, const float4 &x) {
```
what's the motivation for making `a` a reference here? Shouldn't passing by value be perfectly fine here?

mathiaswagner commented on include/float_vector.h:
```diff
@@ -77,14 +84,21 @@ namespace quda {
   }
 
-  __host__ __device__ inline float4 operator+(const float4 x, const float4 y) {
+  __host__ __device__ inline float4 operator+(const float4 &x, const float4 y) {
```
why is x by-reference and y by-value?

mathiaswagner reviewed (approved):
Some minor comments but everything else looks good. Pending some actual testing.

maddyscientist committed:
Optimize getNeigborIndexCB function through reusing earlier full index computation.  Added Coord struct for passing around computed thread coordinates

maddyscientist committed:
Revert "Rework of vector_load/vector_store helpers to allow for potential optimization of indexing"

This reverts commit f1038917a647c0e608c267a57f3ec76aa6915af4.

maddyscientist committed:
Clean up color_spinor.h to use complex i_ helper function for cleaner multpilication by i.  For now use prior wasteful implementation since this regresses performance in single precision on Volta.

maddyscientist committed:
Optimize wilson dslash - fifth dimension index should always be zero

maddyscientist committed:
Apply clang format

maddyscientist committed:
Address review comments

maddyscientist commented:
This is now ready for merging.  I have addressed @mathiaswagner's comments.  Moreover, I have reverted a couple of the optimizations as these were found to regress performance on Volta.  With the pull request as is, I will investigating that for another day in another pull.

maddyscientist commented:
I should note that one specific rework I did was in b4cf12e566a12011c82eff055050399760e9d395, which created a `Coord` object for easier reuse of indexing computations and simplifies the stencil kernels.  This allows for an optimization removing the need to recompute the full lattice index in the neighbour index computation.

mathiaswagner commented on include/dslash_helper.cuh:
```diff
@@ -85,28 +84,28 @@ namespace quda
   */
   template <QudaPCType pc_type, KernelType kernel_type, typename Arg, int nface_ = 1>
-  __host__ __device__ inline int getCoords(int coord[], const Arg &arg, int &idx, int parity, int &dim)
+  __host__ __device__ inline auto getCoords(const Arg &arg, int &idx, int s, int parity, int &dim)
```
love auto here

mathiaswagner reviewed (approved):
Looks good to me
